enhancements/ingress/dynamic-config-manager.md

```diff
@@ -0,0 +1,469 @@
+should be completely transparent to the end-user.  The only effect should be
+that OpenShift router uses less CPU and memory, balances traffic more evenly
+after endpoints updates, and tracks metrics more accurately after route and
+endpoints updates.
+
+#### Variation and form factor considerations [optional]
+
+DCM should function the same on standalone OpenShift, MicroShift, and
+HyperShift.
+
+### API Extensions
+
+DCM does not require any API extensions beyond the featuregate.  However, it
+does have an unsupported config override in the IngressController API in case a
+critical issue is discovered after DCM has been enabled by default.
+
+### Implementation Details/Notes/Constraints [optional]
+
+DCM was implemented in OpenShift 3.  This initial implementation requires
+pre-allocating both backends and servers in order to accommodate routes and
+endpoints that are created after HAProxy starts.
+
+Backends are pre-allocated based on *blueprint routes*, which are route objects
+that the cluster-admin must create in a designated namespace.  A blueprint route
+must specify the TLS termination type and set of annotations.  Backends and
+servers are pre-allocated when HAProxy starts.
+
+If DCM cannot configure something dynamically, it falls back to the old
+fork-and-reload procedure.  In particular, the following conditions require
+falling back to fork and reload:
+
+- More routes are created than the number of pre-allocated backends.
+- More endpoints are created for a route than the number of pre-allocated server slots for that specific route's backend.
+- A route is created that does not match the TLS termination type and annotations of any blueprint route.
+
+In OpenShift 3, configuring blueprint routes was left up to the cluster-admin to
+do.  As such, blueprint routes constitute a user-facing API.  Additionally, the
+cluster-admin could configure the number of server slots to pre-allocate for
+each backend.
+
+Note that pre-allocated server slots and backends take up memory.  In a cluster
+with many routes, pre-allocating multiple server slots for each backend can take
+a considerably large amount of memory.  See
+https://gist.github.com/frobware/2b527ce3f040797909eff482a4776e0b for an
+analysis of the potential memory impact for different choices of balancing
+algorithm, number of threads, number of backends, and number of pre-allocated
+server slots per backend.
+
+To avoid operational overhead, avoid adding new APIs, and minimize overhead of
+enabling DCM, we will initially omit blueprint routes and only pre-allocate 1
+server slot per backend.  This makes DCM an implementation detail; ideally it is
+completely invisible to the cluster-admin, except that the router will fork
+fewer processes and have more accurate metrics and balancing.  Later, we can
+enhance DCM by using newer capabilities in HAProxy 2.y to add arbitrarily many
+servers dynamically, add backends dynamically, and update certificates
+dynamically.
+
```

Comment:
Isn't there anything DCM-related that we could add via the haproxy config template that might require an openshift/API change?  Getting an API framework in place might be useful.   

Are all the Unix socket commands listed in https://docs.haproxy.org/2.8/management.html#9.3 considered to be dynamic config options?  So, not only can we add/del servers, but we could also add/del acls, maps, etc, and disable frontends, servers , healthchecks, and set maxconns/rate-limit, etc.

Reply:
> Isn't there anything DCM-related that we could add via the haproxy config template that might require an openshift/API change? Getting an API framework in place might be useful.

Ideally not.  We want DCM to be as transparent and dynamic as possible, requiring no ahead-of-time configuration.  

> Are all the Unix socket commands listed in https://docs.haproxy.org/2.8/management.html#9.3 considered to be dynamic config options? So, not only can we add/del servers, but we could also add/del acls, maps, etc, and disable frontends, servers , healthchecks, and set maxconns/rate-limit, etc.

Right, any configuration that the router performs using the Unix domain socket is dynamic configuration.  We might indeed be able to use additional capabilities to avoid reloads in more cases.  

Of course, we do need to do some cost-benefit analysis.  It may be worth the engineering effort to implement dynamic configuration of, say certificates but not worth the effort to implement dynamic configuration of rate limiting, depending on how much effort is required and how often people actually update that particular kind of configuration.